pkg/kubelet/pod/pod_manager.go

```diff
@@ -324,3 +328,15 @@ func (pm *basicManager) GetPodByMirrorPod(mirrorPod *v1.Pod) (*v1.Pod, bool) {
 	pod, ok := pm.podByFullName[kubecontainer.GetPodFullName(mirrorPod)]
 	return pod, ok
 }
+
+func (pm *basicManager) GetPodAndMirrorPod(aPod *v1.Pod) (pod, mirrorPod *v1.Pod, wasMirror bool) {
+	pm.lock.RLock()
+	defer pm.lock.RUnlock()
+
+	fullName := kubecontainer.GetPodFullName(aPod)
+	if kubetypes.IsMirrorPod(aPod) {
+		return pm.podByFullName[fullName], aPod, true
+	}
+	return aPod, pm.mirrorPodByFullName[fullName], false
```

Comment:
So pods can be regular or mirror.  You pass an unknown pod type here, and you get back either pod, mirror pod, or both.  From a Kube state diagram:

```
 +->  API --- (spec+mirror) -->-- config loop -->-- pod manager -->-- admission -->-- pod worker
 |                            |                                                           |
 |   File -- (spec) ----------+                                                           |
 |                                                                                        |
 +-------<--------------<------------------ status manager <------------------------------+
```

Mirror pods feed in to API, which means config loop has both, and then before admission we have to separate out the mirror pods and the spec pods.

So the method here takes either, and finds the ones we know about, and only "no pod" is stops us from going to pod manager.

Reply:
Got it, that makes sense. 

I was referring that we already check if `aPod` is a mirror pod, so if it is **not** a mirror pod, I would assume then that `pm.mirrorPodByFullName[fullName]` would be empty. As a result, I was thinking it might be clearer to do `return aPod, nil, false`

Reply:
Hrm, we want to return the mirror pod in that case though, so that it can be passed in to the pod worker.

There's an argument that mirror pod management is a sub responsibility of pod worker (sync methods), not pod manager.  We still need to propagate the config loop items into pod worker.  I think the argument is that the "config -> actual" transition is really a set of controller methods that react to incoming events and notify the right components (mostly pod worker, but maybe others eventually), with a resync method that periodically verifies no delta was missed.  That might be moving all the HandlePod* methods out of kubelet and into a type that calls back to kubelet for specific stuff, but mostly deals with pod worker and a few subsystems.  Maybe not high priority though.

Reply:
We need the mirror pod from this method in all cases.  Maybe I misunderstand the question.

Reply:
ah, my mistake, I read this over again and it makes sense. The key point is that when the argument `aPod` is passed in as not a mirror pod (i.e. regular pod), this regular pod may or may not have a mirror pod, so ` pm.mirrorPodByFullName[fullName]` may or may not be nil depending if a mirror pod exists for it. 

So as you mentioned, we need to return mirror pod in all cases.